File: frontends/telegram/handlers/messages.py
"""Text message handlers for the Notes Bot."""

import logging
from telegram import Update
from telegram.ext import ContextTypes

from ..config import ROOT_ID
from ..grpc_client import core_client
from ..states.context import UserState
from ..states import state_manager
from ..keyboards.main_menu import get_main_menu_keyboard
from ..middleware import reply_message
from ..utils import escape_markdown_v2
from .reminders import (
    handle_reminder_title_input,
    handle_reminder_param_input,
)

logger = logging.getLogger(__name__)


async def handle_text_message(
    update: Update, context: ContextTypes.DEFAULT_TYPE
) -> None:
    if not update.message or not update.effective_user or not update.message.text:
        return

    user_id = update.effective_user.id
    text = update.message.text.strip()

    if ROOT_ID and user_id != ROOT_ID:
        await reply_message(
            update_or_query=update,
            text="⛔ Unauthorized access.",
        )
        logger.warning(f"Unauthorized message from user {user_id}")
        return

    user_context = state_manager.get_context(user_id)
    current_state = user_context.state
    active_date = user_context.active_date

    try:
        if current_state == UserState.WAITING_RATING:
            try:
                rating = int(text)
                if rating < 0 or rating > 10:
                    await reply_message(
                        update_or_query=update,
                        text="❌ Оценка должна быть от 0 до 10\\. Попробуйте снова\\.",
                    )
                    return

                if core_client.update_rating(active_date, rating):
                    state_manager.update_context(user_id, state=UserState.IDLE)
                    await reply_message(
                        update_or_query=update,
                        text=f"✅ Оценка {rating} сохранена\\!",
                        keyboard=get_main_menu_keyboard(active_date),
                    )
                    logger.info(f"User {user_id} set rating {rating} for {active_date}")
                else:
                    await reply_message(
                        update_or_query=update,
                        text="❌ Ошибка при сохранении оценки\\.",
                    )

            except ValueError:
                await reply_message(
                    update_or_query=update,
                    text="❌ Пожалуйста, введите число от 0 до 10\\.",
                )

        elif current_state == UserState.REMINDER_CREATE_TITLE:
            await handle_reminder_title_input(update, user_id, text)

        elif current_state in (
            UserState.REMINDER_CREATE_TIME,
            UserState.REMINDER_CREATE_DAY,
            UserState.REMINDER_CREATE_INTERVAL,
            UserState.REMINDER_CREATE_DATE,
        ):
            await handle_reminder_param_input(update, user_id, text)

        elif current_state == UserState.WAITING_NEW_TASK:
            if core_client.add_task(active_date, text):
                state_manager.update_context(user_id, state=UserState.TASKS_VIEW)
                await reply_message(
                    update_or_query=update,
                    text=f"✅ Задача добавлена: {escape_markdown_v2(text)}",
                )
                logger.info(f"User {user_id} added task: {text}")
                await reply_message(
                    update_or_query=update,
                    text='Используйте кнопку "Задачи" для просмотра\\.',
                    keyboard=get_main_menu_keyboard(active_date),
                )
            else:
                await reply_message(
                    update_or_query=update,
                    text="❌ Ошибка при добавлении задачи\\.",
                )

        else:
            if core_client.append_to_note(active_date, text):
                await reply_message(
                    update_or_query=update,
                    text=f"✅ Текст добавлен в заметку {escape_markdown_v2(active_date)}",
                    keyboard=get_main_menu_keyboard(active_date),
                )
                logger.info(f"User {user_id} added text to {active_date}")
            else:
                await reply_message(
                    update_or_query=update,
                    text="❌ Ошибка при сохранении текста\\.",
                )

    except Exception as e:
        logger.error(f"Error handling text message from user {user_id}: {e}")
        await reply_message(
            update_or_query=update,
            text="❌ Произошла ошибка при обработке сообщения\\.",
        )

File: frontends/telegram/tghandlers/messages.go
package tghandlers

import (
	"context"
	"fmt"
	"runtime/debug"
	"strconv"
	"strings"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
	"go.opentelemetry.io/otel/attribute"
	"go.uber.org/zap"

	"notes_bot/frontends/telegram/tgstates"
	"notes_bot/internal/applog"
	"notes_bot/internal/telemetry"
)

type stateTextHandler func(a *App, ctx context.Context, tgBot *tgbotapi.BotAPI, update *tgbotapi.Update, chatID, userID int64, text string, uc *tgstates.UserContext)

var stateTextHandlers = map[tgstates.UserState]stateTextHandler{
	tgstates.StateWaitingRating: func(a *App, ctx context.Context, tgBot *tgbotapi.BotAPI, _ *tgbotapi.Update, chatID, userID int64, text string, uc *tgstates.UserContext) {
		a.handleRatingInput(ctx, tgBot, chatID, userID, text, uc.ActiveDate)
	},
	tgstates.StateReminderCreateTitle: func(a *App, ctx context.Context, tgBot *tgbotapi.BotAPI, update *tgbotapi.Update, _, userID int64, text string, _ *tgstates.UserContext) {
		a.handleReminderTitleInput(ctx, tgBot, update, userID, text)
	},
	tgstates.StateReminderCreateTime: func(a *App, ctx context.Context, tgBot *tgbotapi.BotAPI, update *tgbotapi.Update, _, userID int64, text string, _ *tgstates.UserContext) {
		a.handleReminderParamInput(ctx, tgBot, update, userID, text)
	},
	tgstates.StateReminderCreateDay: func(a *App, ctx context.Context, tgBot *tgbotapi.BotAPI, update *tgbotapi.Update, _, userID int64, text string, _ *tgstates.UserContext) {
		a.handleReminderParamInput(ctx, tgBot, update, userID, text)
	},
	tgstates.StateReminderCreateInterval: func(a *App, ctx context.Context, tgBot *tgbotapi.BotAPI, update *tgbotapi.Update, _, userID int64, text string, _ *tgstates.UserContext) {
		a.handleReminderParamInput(ctx, tgBot, update, userID, text)
	},
	tgstates.StateReminderCreateDate: func(a *App, ctx context.Context, tgBot *tgbotapi.BotAPI, update *tgbotapi.Update, _, userID int64, text string, _ *tgstates.UserContext) {
		a.handleReminderParamInput(ctx, tgBot, update, userID, text)
	},
	tgstates.StateWaitingNewTask: func(a *App, ctx context.Context, tgBot *tgbotapi.BotAPI, _ *tgbotapi.Update, chatID, userID int64, text string, uc *tgstates.UserContext) {
		a.handleAddTaskInput(ctx, tgBot, chatID, userID, text, uc.ActiveDate)
	},
}

func (a *App) HandleTextMessage(ctx context.Context, tgBot *tgbotapi.BotAPI, update *tgbotapi.Update) {
	if update.Message == nil || update.Message.From == nil || update.Message.Text == "" {
		return
	}

	ctx, span := telemetry.StartSpan(ctx)
	defer span.End()

	log := applog.With(ctx, a.Logger)
	userID := update.Message.From.ID
	if !a.authorized(userID) {
		sendText(ctx, tgBot, update.Message.Chat.ID, "⛔ Unauthorized access.", nil, true)
		log.Warn("unauthorized message", zap.Int64("user_id", userID))
		return
	}

	text := strings.TrimSpace(update.Message.Text)
	uc, err := a.State.GetContext(ctx, userID)
	if err != nil {
		log.Error("get context", zap.Error(err))
		return
	}
	span.SetAttributes(attribute.String("user.state", string(uc.State)))

	chatID := update.Message.Chat.ID

	defer func() {
		if r := recover(); r != nil {
			log.Error("panic in text handler", zap.Any("recover", r), zap.String("stack", string(debug.Stack())))
		}
	}()

	if h, ok := stateTextHandlers[uc.State]; ok {
		h(a, ctx, tgBot, update, chatID, userID, text, uc)
	} else {
		a.handleAppendNote(ctx, tgBot, chatID, userID, text, uc.ActiveDate)
	}
}

func (a *App) handleRatingInput(ctx context.Context, tgBot *tgbotapi.BotAPI, chatID, userID int64, text, activeDate string) {
	ctx, span := telemetry.StartSpan(ctx, attribute.String("note.date", activeDate))
	defer span.End()

	log := applog.With(ctx, a.Logger)
	rating, err := strconv.Atoi(text)
	if err != nil || rating < 0 || rating > 10 {
		sendText(ctx, tgBot, chatID, "❌ Оценка должна быть от 0 до 10. Попробуйте снова.", nil, true)
		return
	}

	ok, err := a.Core.UpdateRating(ctx, activeDate, rating)
	if err != nil || !ok {
		sendText(ctx, tgBot, chatID, "❌ Ошибка при сохранении оценки.", nil, true)
		return
	}

	a.State.UpdateContext(ctx, userID, func(uc *tgstates.UserContext) {
		uc.State = tgstates.StateIdle
	})
	kb := a.getMainMenuKeyboard(ctx)
	sendText(ctx, tgBot, chatID, fmt.Sprintf("✅ Оценка %d сохранена!", rating), &kb, true)
	log.Info("user set rating", zap.Int64("user_id", userID), zap.Int("rating", rating))
}

func (a *App) handleAddTaskInput(ctx context.Context, tgBot *tgbotapi.BotAPI, chatID, userID int64, text, activeDate string) {
	ctx, span := telemetry.StartSpan(ctx, attribute.String("note.date", activeDate))
	defer span.End()

	log := applog.With(ctx, a.Logger)
	ok, err := a.Core.AddTask(ctx, activeDate, text)
	if err != nil || !ok {
		sendText(ctx, tgBot, chatID, "❌ Ошибка при добавлении задачи.", nil, true)
		return
	}
	a.State.UpdateContext(ctx, userID, func(uc *tgstates.UserContext) {
		uc.State = tgstates.StateTasksView
	})
	sendText(ctx, tgBot, chatID, fmt.Sprintf("✅ Задача добавлена: %s", text), nil, true)
	kb := a.getMainMenuKeyboard(ctx)
	sendText(ctx, tgBot, chatID, "Используйте кнопку \"Задачи\" для просмотра.", &kb, true)
	log.Info("user added task", zap.Int64("user_id", userID))
}

func (a *App) handleAppendNote(ctx context.Context, tgBot *tgbotapi.BotAPI, chatID, userID int64, text, activeDate string) {
	ctx, span := telemetry.StartSpan(ctx, attribute.String("note.date", activeDate))
	defer span.End()

	log := applog.With(ctx, a.Logger)
	ok, err := a.Core.AppendToNote(ctx, activeDate, text)
	if err != nil || !ok {
		sendText(ctx, tgBot, chatID, "❌ Ошибка при сохранении текста.", nil, true)
		return
	}
	kb := a.getMainMenuKeyboard(ctx)
	sendText(ctx, tgBot, chatID, fmt.Sprintf("✅ Текст добавлен в заметку %s", activeDate), &kb, true)
	log.Info("user appended text", zap.Int64("user_id", userID))
}
